services/webnn/ort/graph_builder_ort.cc

```diff
@@ -952,6 +986,123 @@ GraphBuilderOrt::AddExpandOperation(const mojom::Expand& expand) {
   return base::ok();
 }
 
+[[nodiscard]] base::expected<void, mojom::ErrorPtr>
+GraphBuilderOrt::AddDequantizeLinearOperation(
+    const mojom::DequantizeLinear& dequantize_linear) {
+  const std::string node_name =
+      GenerateNextOperationName(dequantize_linear.label);
+  std::string input_name =
+      GetOperandNameById(dequantize_linear.input_operand_id);
+  std::string scale_name =
+      GetOperandNameById(dequantize_linear.scale_operand_id);
+  std::string zero_point_name =
+      GetOperandNameById(dequantize_linear.zero_point_operand_id);
+  std::string output_name =
+      GetOperandNameById(dequantize_linear.output_operand_id);
+
+  const OperandDescriptor& input_descriptor =
+      GetOperand(dequantize_linear.input_operand_id).descriptor;
+  std::vector<uint32_t> input_shape = input_descriptor.shape();
+
+  const OperandDescriptor& scale_descriptor =
+      GetOperand(dequantize_linear.scale_operand_id).descriptor;
+  std::vector<uint32_t> scale_shape = scale_descriptor.shape();
+
+  int64_t axis = 1;
+  int64_t block_size = 0;
+  bool need_transpose = false;
+
+  // https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/dequantize_linear.cpp#L220
+  if (scale_shape.size() > 2) {
+    return NewNotSupportedError(
+        "OpenVINO dequantizeLinear cannot operate with more than 2D scales");
+  }
+
+  if (scale_shape.empty()) {
+    // For per-tensor/layer dequantization the scale is a scalar.
+    axis = 0;
+  } else if (scale_shape.size() == 1) {
+    bool is_valid = false;
+    // for per per-axis dequantization it is a 1-D Tensor
+    for (size_t i = 0; i < input_shape.size(); i++) {
+      if (scale_shape[0] == input_shape[i]) {
+        axis = i;
+        is_valid = true;
+      }
+    }
+    if (!is_valid) {
+      return NewNotSupportedError(
+          "For 1D scale, the size of scale must be the same as the size of the "
+          "input dim specified by the axis.");
+    }
+  } else {
+    CHECK_EQ(scale_shape.size(), 2u);
+    // For blocked dequantization it has the same shape as the input, except for
+    // one dimension in which blocking is performed.
+    if (scale_shape.size() == input_shape.size()) {
+      uint32_t diff_count = 0;
+      for (size_t i = 0; i < input_shape.size(); i++) {
+        if (scale_shape[i] != input_shape[i]) {
+          // https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/dequantize_linear.cpp#L230
+          if (input_shape[i] % scale_shape[i] != 0) {
+            return NewNotSupportedError(
+                "For blocked dequantization, OpenVINO DequantizeLinear doesn't "
+                "support case when input cannot be divided by scale.");
+          }
+          block_size = input_shape[i] / scale_shape[i];
+          axis = i;
+          diff_count++;
+          if (diff_count > 1) {
```

Comment:
Are there any plan to support multiple blocked axes? Add a TODO?

Reply:
Sounds good. I will add a TODO here.